myst_nbs/case_studies/wrapping_jax_function.myst.md

```diff
@@ -0,0 +1,765 @@
+]
+
+az.plot_posterior(idata, ref_val=true_values, grid=(3, 5));
+```
+
+The posteriors look reasonably centered around the true values used to generate our data.
+
++++
+
+## Unwrapping the wrapped JAX function
+
++++
+
+As mentioned in the beginning, Aesara can compile an entire graph to JAX. To do this, it needs to know how each {class}`~aesara.graph.op.Op` in the graph can be converted to a JAX function. This can be done by {term}`dispatch <dispatching>` with {func}`aesara.link.jax.dispatch.jax_funcify`. Most of the default Aesara {class}`~aesara.graph.op.Op`s already have such a dispatch function, but we will need to add a new one for our custom `HMMLogpOp`, as Aesara has never seen that before.
+
+For that we need a function which returns (another) JAX function, that performs the same computation as in our `perform` method. Fortunately, we started exactly with such function, so this amounts to 3 short lines of code.
+
+```{code-cell} ipython3
+@jax_funcify.register(HMMLogpOp)
+def hmm_logp_dispatch(op, **kwargs):
+    return vec_hmm_logp
+```
+
+:::{note}
+We do not return the jitted function, so that the entire Aesara graph can be jitted together after being converted to JAX.
+:::
+
+For a better understanding of {class}`~aesara.graph.op.Op` JAX conversions, we recommend reading Aesara's [Adding JAX and Numba support for Ops guide](https://aesara.readthedocs.io/en/latest/extending/creating_a_numba_jax_op.html?highlight=JAX)
+
+We can test that our conversion function is working properly by compiling a {func}`aesara.function` with `mode="JAX"`:
+
+```{code-cell} ipython3
+out = hmm_logp_op(
+    emission_observed,
+    emission_signal_true,
+    emission_noise_true,
+    logp_initial_state_true,
+    logp_transition_true,
+)
+jax_fn = aesara.function(inputs=[], outputs=out, mode="JAX")
+jax_fn()
+```
+
+We can also compile a JAX function that computes the log probability of each variable in our PyMC model, similar to {meth}`~pymc.Model.point_logps`. We will use the helper method {meth}`~pymc.Model.compile_fn`.
+
+```{code-cell} ipython3
+model_logp_jax_fn = model.compile_fn(model.logpt(sum=False), mode="JAX")
+model_logp_jax_fn(initial_point)
+```
+
+Note that we could have added an equally simple function to convert our `HMMLogpGradOp`, in case we wanted to convert Aesara gradient graphs to JAX. In our case, we don't need to do this because we will rely on JAX `grad` function (or more precisely, NumPyro will rely on it) to obtain these again from our compiled JAX function.
+
+We include a {ref}`short discussion <aesara_vs_jax>` at the end of this document, to help you better understand the trade-offs between working with Aesara graphs vs JAX functions, and when you might want to use one or the other.
+
++++
+
+### Sampling with NumPyro
+
++++
+
+Now that we know our model logp can be entirely compiled to JAX, we can use the handy {mod}`pymc.sampling_jax` module to sample our model using the pure JAX sampler implemented in NumPyro.
```

Comment:
`pymc.sampling_jax` is not documented as a module, instead its functions are integrated in the sampling section with the "regular" ones: https://docs.pymc.io/en/latest/api/samplers.html#samplers. Should it be documented as a module?

Reply:
Probably not, but we should add a note saying that the module has to be imported to be able to call those functions

Reply:
Opened an issue here: https://github.com/pymc-devs/pymc/issues/5710